Objects/listobject.c

```diff
@@ -2682,16 +2685,34 @@ list_index_impl(PyListObject *self, PyObject *value, Py_ssize_t start,
         if (stop < 0)
             stop = 0;
     }
-    for (i = start; i < stop && i < Py_SIZE(self); i++) {
-        PyObject *obj = self->ob_item[i];
-        Py_INCREF(obj);
-        int cmp = PyObject_RichCompareBool(obj, value, Py_EQ);
-        Py_DECREF(obj);
-        if (cmp > 0)
-            return PyLong_FromSsize_t(i);
-        else if (cmp < 0)
-            return NULL;
-    }
+    if (!key)
+        for (i = start; i < stop && i < Py_SIZE(self); i++) {
+            PyObject *obj = self->ob_item[i];
+            /* take reference to item, in case comparison methods modify list */
+            Py_INCREF(obj);
+            int cmp = PyObject_RichCompareBool(obj, value, Py_EQ);
+            Py_DECREF(obj);
+            if (cmp > 0)
+                return PyLong_FromSsize_t(i);
+            else if (cmp < 0)
+                return NULL;
+        }
+    else
+        for (i = start; i < stop && i < Py_SIZE(self); i++) {
+            PyObject *item = self->ob_item[i];
+            /* take reference to item, in case key modifies list */
+            Py_INCREF(item);
+            PyObject *obj = PyObject_CallFunctionObjArgs(key, item, NULL);
+            Py_DECREF(item);
+            if (!obj)
```

Comment:
PEP 7

Reply:
would be good if you  pointed to the particular pep7 violation you have in mind.   I don't think constructs such as `!obj` are forbidden, if that is what you mean.